sql-plugin/src/main/scala/org/apache/spark/sql/rapids/RapidsShuffleInternalManagerBase.scala

```diff
@@ -63,12 +73,342 @@ abstract class GpuShuffleBlockResolverBase(
+   * and then call stop() with success = false if they get an exception, we want to make sure
+   * we don't try deleting files, etc twice.
+   */
+  private var stopping = false
+
+  val diskBlockObjectWriters = new mutable.HashMap[Int, (Int, DiskBlockObjectWriter)]()
+
+  override def write(records: Iterator[Product2[K, V]]): Unit = {
+    withResource(new NvtxRange("ThreadedWriter.write", NvtxColor.RED)) { _ =>
+      withResource(new NvtxRange("compute", NvtxColor.GREEN)) { _ =>
+        var errorOcurred: Throwable = null
+        val mapOutputWriter = shuffleExecutorComponents.createMapOutputWriter(
+          shuffleId,
+          mapId,
+          numPartitions)
+        try {
+          val partLengths = if (!records.hasNext) {
+            commitAllPartitions(mapOutputWriter, true /*empty checksum*/)
+          } else {
+            // per reduce partition id
+            // open all the writers ahead of time (Spark does this already)
+            val openStartTime = System.nanoTime()
+            (0 until numPartitions).map { i =>
+              val (blockId, file) = blockManager.diskBlockManager.createTempShuffleBlock()
+              val writer: DiskBlockObjectWriter = blockManager.getDiskWriter(
+                blockId, file, serializer, fileBufferSize, writeMetrics)
+              setChecksumIfNeeded(writer, i) // spark3.2.0+
+
+              // Places writer objects at round robin slot numbers apriori
+              // this choice is for simplicity but likely needs to change so that
+              // we can handle skew better
+              val slotNum = RapidsShuffleInternalManagerBase.getNextSlot
+              diskBlockObjectWriters.put(i, (slotNum, writer))
+            }
+            writeMetrics.incWriteTime(System.nanoTime() - openStartTime);
+
+            // we call write on every writer for every record in parallel
+            val writeFutures = new ArrayBuffer[Future[Unit]]
+            records.foreach { record =>
+              val key = record._1
+              val value = record._2
+              val reducePartitionId: Int = partitioner.getPartition(key)
+              val (slotNum, myWriter) = diskBlockObjectWriters(reducePartitionId)
+
+              // we close batches actively in the `records` iterator as we get the next batch
+              // this makes sure it is kept alive while a task is able to handle it.
+              val cb = value match {
+                case columnarBatch: ColumnarBatch =>
+                  SlicedGpuColumnVector.incRefCount(columnarBatch)
+                case _ =>
+                  null
+              }
+
+              writeFutures += RapidsShuffleInternalManagerBase.queueTask(slotNum, () => {
+                withResource(cb) { _ =>
+                  myWriter.write(key, value)
+                }
+              })
+            }
+
+            withResource(new NvtxRange("WaitingForWrites", NvtxColor.PURPLE)) { _ =>
+              synchronized {
```

Comment:
Is this synchronized block still needed, or is this an unnecessary remnant from when it was using wait/notify? We're walking a local `writeFutures` here.